internal/main.go

```diff
@@ -68,12 +92,20 @@ func Run(logger *slog.Logger, criSocketPath string, dockerConfigJSONPath string,
 				},
 				Auth: auth,
 			}
-			go pullImageWithRetries(ctx, logger.With("image", imageName, "authNum", i), &wg, criClient, metricsSink.Chan(), imageName, request, timing)
+			go pullImageWithRetries(ctx, logger.With("image", imageName, "authNum", i), &wg, criClient, metricsSink.Chan(), imageName, request, timing, &results)
 		}
 	}
 	wg.Wait()
 	logger.Info("pulling images finished")
 	metricsSink.Await()
+
+	if k8sClient != nil {
+		// Don't fail the overall operation if node labeling fails.
+		if err := nodelabels.UpdateNodeLabels(ctx, k8sClient, nodeName, instanceName, &results, logger); err != nil {
+			logger.Error("failed to update node labels", "error", err)
+		}
+	}
+
```

Comment:
...and this block into a new function and call it from `Run` so that it does not turn into spaghetti. 😉